Update Object.wait to use Blocker.begin/end for Java 19+

fengxue-IS opened the pull request:
Fix: #16198 

Signed-off-by: Jack Lu <Jack.S.Lu@ibm.com>

fengxue-IS commented:
Personal build https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/14803/

@tajila can you please take a look
@0xdaryl can you review the JIT changes please, I only found 1 instance where `Object.wait` is hardcoded in JIT

0xdaryl commented on runtime/compiler/codegen/J9CodeGenerator.cpp:
```diff
@@ -532,7 +532,7 @@ static bool scanForNativeMethodsUntilMonitorNode(TR::TreeTop *firstTree, TR::Com
          TR::ResolvedMethodSymbol *symbol = callTestNode->getSymbol()->castToResolvedMethodSymbol();
          if (strstr(symbol->signature(comp->trMemory()), "java/lang/Object.notify") ||
-             strstr(symbol->signature(comp->trMemory()), "java/lang/Object.wait"))
+             strstr(symbol->signature(comp->trMemory()), "java/lang/Object.waitImpl"))
```
I wonder if the check needs to be for `Object.wait` OR `Object.waitImpl` to be consistent with the original semantics when `Object.wait` was native.  If `Object.wait` is not inlined by the JIT, the semantics are the same as before.  But if `Object.wait` is inlined then we want the check to be for the native `Object.waitImpl` method.

fengxue-IS replied:
based on the code that sets `callTestNode`, it is checking for resolved native node, so the original code is definitely looking for the native method, I don't know if there is special handling on the JIT side where `Object.wait` is recognized and inlined differently than normal methods (I wasn't able to find such code in OpenJ9 though)

0xdaryl replied:
I'm pasting this here because I'm tracing the history of this code to get the context.  It was a performance feature introduced early days in lock reservation implementation, motivated thusly:
```
This is a problem with excessive cancellations happening while
doing notify and wait on an object that was reserved. We shouldn't 
reserve objects that we do notify or wait on
```
We'll have to make sure that the semantics are preserved.  Still looking into it...

gacholio replied:
I think the `strstr` would match the new native without changing the code here. It's not matching signatures here, so it's trying to catch all of the wait methods in one check.

Also, `waitImpl` is private, so it will never be called directly, only from the other `wait` implemenatations.

I suggest removing this part of the change entirely.

gacholio replied:
In fact, this change is clearly incorrect given the matching using `strstr`. It absolutely has to be removed.

fengxue-IS replied:
fixed

fengxue-IS commented:
> I think it might be best to add `waitImpl` to the final method list and leave `wait` in there - these are methods that do not require vTable entries (so we would save one slot in every class). Please give it a try and see if it causes any problems.

I've modified final.c to also include `waitImpl` and increased the `OBJECT_FINAL_COUNT`.
Did I miss any place that have a separate list?

gacholio commented:
Sorry, I misread the diff - you have done exactly what I was suggesting.

gacholio commented on runtime/vm/BytecodeInterpreter.hpp:
```diff
@@ -4853,7 +4853,7 @@ class INTERPRETER_CLASS
-	/* java.lang.Object: public final native void wait(long millis, int nanos); */
+	/* java.lang.Object: public final native void waitImpl(long millis, int nanos); */
 	VMINLINE VM_BytecodeAction
 	inlObjectWait(REGISTER_ARGS_LIST)
```
Should really change the method name to match the native name (no functional effect, just consistency).

fengxue-IS replied:
updated INL name to `inlObjectWaitImpl`

gacholio commented:
Is this the approach hotspot took? I'm concerned that people know too much about `wait` being native and this may cause issues with instrumentation agents.

fengxue-IS commented:
RI have a similar approach to add `Object.wait0`
Even without this change, RI uses `native wait(long milli)` as the base where OpenJ9 uses `native wait(long milli, int nano)` as base.

gacholio commented:
> RI uses `native wait(long milli)` as the base

Does this mean they don't support nanos (ignore it and forward to the above native) or are there multiple natives?

fengxue-IS commented:
> > RI uses `native wait(long milli)` as the base
> 
> Does this mean they don't support nanos (ignore it and forward to the above native) or are there multiple natives?

They round nanos to millis

gacholio commented:
> They round nanos to millis

Perhaps we should do the same (not in this PR).

Given the existing differences between the implementations, I believe instrumentation agents are unlikely to be affected by this change.

fengxue-IS commented:
Addressed review comments, local testing passed on xlinux, personal build in progress: https://hyc-runtimes-jenkins.swg-devops.com/view/OpenJ9%20-%20Personal/job/Pipeline-Build-Test-Personal/14854/

gacholio commented:
jenkins test sanity zlinux jdk8,jdk19

gacholio commented:
jenkins compile win jdk19
